source/agora/consensus/validation/Transaction.d

```diff
@@ -119,9 +119,22 @@ public string isInvalidReason (
             }
         }
 
-        // current limitation: if any UTXO is frozen, they all must be frozen
-        if ((count_freeze > 0) && (count_freeze != tx.inputs.length))
-            return "Transaction: Rejected combined inputs (freeze & payment)";
+        if (tx.type == TxType.Payment)
+        {
+            // current limitation: if any UTXO is frozen, they all must be frozen
+            if ((count_freeze > 0) && (count_freeze != tx.inputs.length))
+                return "Transaction: Rejected combined inputs (freeze & payment)";
+        }
+        else
+        {
+            if (count_freeze > 0)
+                return "Transaction: Frozen UTXO is used for data storage";
+
+            if (tx.data.is_oversize)
+                return "Transaction: Data is too large (Maximum size is 1024 bytes)";
```

Comment:
This just reads the data sent by the client. If the client make its own binary data and set the flag to false, this will pass.